lib/model/avatar_url.dart

```diff
@@ -0,0 +1,148 @@
+import 'dart:convert';
+
+import 'package:crypto/crypto.dart';
+import 'package:path/path.dart';
+
+import '../log.dart';
+
+/// Pixel dimensions of different size choices we have (they're all
+/// square) when requesting an uploaded avatar.
+/// DEFAULT_AVATAR_SIZE in zerver/lib/upload.py.
+const int defaultUploadSizePx = 100;
+
+/// MEDIUM_AVATAR_SIZE in zerver/lib/upload.py.
+const int mediumUploadSizePx = 500;
+
+abstract class AvatarURL {
+  static AvatarURL fromUserOrBotData({
+    String? rawAvatarUrl,
+    required Uri realm,
+    required int userId,
+    required String email,
+  }) {
+    if (rawAvatarUrl == null || rawAvatarUrl.isEmpty) {
+      return GravatarURL.validateAndConstructInstance(email: email);
+    } else if (rawAvatarUrl.startsWith(GravatarURL.origin)) {
+      return GravatarURL.validateAndConstructInstance(
+          urlFromServer: rawAvatarUrl, email: email);
+    } else if (['.png','.jpg','.jpeg'].contains(extension(rawAvatarUrl))) {
```

Comment:
What's the basis for this condition?

Reply:
I was confused with the implementation in zulip-mobile where the get() method in UploadAvatarUrl would replace .png url with -medium.png. But a few avatar url's in zulip like the bots would not have a .png file extension, hence I thought this would be handled by the FallbackAvatarUrl.

Now I've written a way to handle both cases in the UploadAvatarUrl class itself, as to my understanding FallbackAvatarUrl should only be used when an avatarUrl is not sent by the rest api.
[869bd80](https://github.com/zulip/zulip-flutter/pull/534/commits/869bd8005ee10010db251e20083604358695d9c3)

Reply:
Yeah, FallbackAvatarUrl is only for the case when the server doesn't tell us where the avatar is.

It looks like your new version still has special logic for `.jpg` and `.jpeg` extensions. That doesn't correspond to anything in the zulip-mobile code; the logic there around adding "-medium" is:
```js
      result.pathname = result.pathname.replace(
        // `.png` is optional: s3 uploads don't use it, local ones do.
        // See TODO in zulip/zulip@c03615b98.
        /(?:\.png)?$/,
        '-medium.png',
      );
```

For most things where we're interpreting data from the server or deciding what requests to make to the server, we follow the Zulip API docs as a specification for how to interact with the server. Here, as @chrisbobbe said in the #301 description:
> On CZO in `#api documentation`, I've [summarized](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/avatar_url.3A.20Request.20specific.20size/near/1636546) zulip-mobile's logic for determining a URL for any given size in physical pixels. I'm hoping the API doc can start specifying that logic, and we can use the new documentation for zulip-flutter's implementation.

So these avatar URLs are a part of the API that isn't yet documented, but we should do the next best thing to following the API docs:
* follow @chrisbobbe's summary in that chat thread, which is effectively a draft of what the API docs should say;
* follow the behavior of the zulip-mobile implementation, which encodes what we've previously learned about the server's behavior (and is what @chrisbobbe drew on for that chat thread).

Reply:
[f4bd2a9](https://github.com/zulip/zulip-flutter/pull/534/commits/f4bd2a94630b52f45ada6ed0ef6202766efb31cb)